networkit/cpp/community/ParallelLeiden.cpp

```diff
@@ -0,0 +1,530 @@
+            if (!queue.empty()) {
+                std::swap(currentNodes, queue.front());
+                queue.pop();
+            } else { // queue empty && newNodes empty
+                waitingForNodes++;
+                if (waitingForNodes < tcount) { // Not all nodes are done yet, wait for new work
+                    waitingForResize++;
+                    while (queue.empty() && waitingForNodes < tcount) {
+                        workAvailable.wait(uniqueLock);
+                    }
+                    if (waitingForNodes < tcount) {
+                        // Notified and not all done means there's new work
+                        std::swap(currentNodes, queue.front());
+                        queue.pop();
+                        waitingForNodes--;
+                        waitingForResize--;
+                        continue;
+                    }
+                }
+                uniqueLock.unlock(); // Notified and all done, stop.
+                workAvailable.notify_all();
+                break;
+            }
+        } while (true);
+        TRACE("Thread " + std::to_string(omp_get_thread_num()) + " worked "
+              + std::to_string(totalNodesPerThread[omp_get_thread_num()]) + "Nodes and moved "
+              + std::to_string(moved[omp_get_thread_num()]));
+    }
+    result.setUpperBound(upperBound);
+    assert(queue.empty());
+    assert(waitingForNodes == tcount);
+    if (Aux::Log::isLogLevelEnabled(Aux::Log::LogLevel::debug)) {
+        count totalMoved = accumulate(moved.begin(), moved.end(), (count)0);
+        count totalWorked =
+            accumulate(totalNodesPerThread.begin(), totalNodesPerThread.end(), (count)0);
+        std::cout << "[DEBUG] Total worked: " + std::to_string(totalWorked)
+                         + " Total moved: " + std::to_string(totalMoved)
+                         + " moved to singleton community: " + std::to_string(singleton)
+                  << std::endl;
+    }
+}
+
+Partition ParallelLeiden::parallelRefine(const Graph &graph) {
+    Partition refined(graph.numberOfNodes());
+    refined.allToSingletons();
+    DEBUG("Starting refinement with " + std::to_string(result.numberOfSubsets()) + " partitions");
+    std::vector<uint_fast8_t> singleton(refined.upperBound(), true);
+    std::vector<double> cutCtoSminusC(refined.upperBound());
+    std::vector<double> refinedVolumes(refined.upperBound()); // Community Volumes P_refined
+    std::vector<std::mutex> locks(refined.upperBound());
+    std::vector<node> nodes(graph.upperNodeIdBound(), none);
+#pragma omp parallel
+    {
+        std::vector<index> neighComms;
+        // Keeps track of relevant Neighbor communities. Needed to reset the clearlist fast
+        std::vector<double> cutWeights(refined.upperBound()); // cut from Node to Communities
+        auto &mt = Aux::Random::getURNG();
+#pragma omp for
+        for (long u = 0; u < static_cast<long>(graph.upperNodeIdBound()); u++) {
```

Comment:
Here you can use the `omp_index` type, it is defined in `Globals.hpp`.

Reply:
The type has to be signed because of openmp restrictions

Reply:
Only if you use MSVC; in that case, `omp_index` is signed.

Reply:
Oh, ok